langchain/agents/agent_toolkits/openapi/planner.py

```diff
@@ -0,0 +1,215 @@
+"""Agent that interacts with OpenAPI APIs via a hierarchical planning approach."""
+import json
+import re
+from typing import List, Optional
+
+import yaml
+
+from langchain.agents.agent import AgentExecutor
+from langchain.agents.agent_toolkits.openapi.planner_prompt import (
+    API_CONTROLLER_PROMPT,
+    API_CONTROLLER_TOOL_DESCRIPTION,
+    API_CONTROLLER_TOOL_NAME,
+    API_ORCHESTRATOR_PROMPT,
+    API_PLANNER_PROMPT,
+    API_PLANNER_TOOL_DESCRIPTION,
+    API_PLANNER_TOOL_NAME,
+    PARSING_GET_PROMPT,
+    PARSING_POST_PROMPT,
+    REQUESTS_GET_TOOL_DESCRIPTION,
+    REQUESTS_POST_TOOL_DESCRIPTION,
+)
+from langchain.agents.agent_toolkits.openapi.spec import (
+    ReducedOpenAPISpec,
+)
+from langchain.agents.mrkl.base import ZeroShotAgent
+from langchain.agents.tools import Tool
+from langchain.chains.llm import LLMChain
+from langchain.llms.openai import OpenAI
+from langchain.prompts import PromptTemplate
+from langchain.requests import RequestsWrapper
+from langchain.schema import BaseLanguageModel
+from langchain.tools.base import BaseTool
+from langchain.tools.requests.tool import BaseRequestsTool
+
+#
+# Requests tools with LLM-instructed extraction of truncated responses.
+#
+# Of course, truncating so bluntly may lose a lot of valuable
+# information in the response.
+# However, the goal for now is to have only a single inference step.
+MAX_RESPONSE_LENGTH = 5000
+
+
+class RequestsGetToolWithParsing(BaseRequestsTool, BaseTool):
+    name = "requests_get"
+    description = REQUESTS_GET_TOOL_DESCRIPTION
+    response_length: Optional[int] = MAX_RESPONSE_LENGTH
+    llm_chain = LLMChain(
+        llm=OpenAI(),
+        prompt=PARSING_GET_PROMPT,
+    )
+
+    def _run(self, text: str) -> str:
+        try:
+            data = json.loads(text)
+        except json.JSONDecodeError as e:
+            raise e
+        response = self.requests_wrapper.get(data["url"])
+        response = response[: self.response_length]
+        return self.llm_chain.predict(
+            response=response, instructions=data["output_instructions"]
+        ).strip()
+
+    async def _arun(self, text: str) -> str:
+        raise NotImplementedError()
+
+
+class RequestsPostToolWithParsing(BaseRequestsTool, BaseTool):
+    name = "requests_post"
+    description = REQUESTS_POST_TOOL_DESCRIPTION
+
+    response_length: Optional[int] = MAX_RESPONSE_LENGTH
+    llm_chain = LLMChain(
+        llm=OpenAI(),
```

Comment:
@jerwelborn Probably late to comment here but wanting to bring this to your attention. Is it possible to not hardcode the llm here and allow user to specify which model to use?

Reply:
Also really awesome job you did here. Similar approach is definitely needed for SQL agent also. Maybe there is a way to generalize the hierarchical planning agent so it can be composed over other agents.